src/memray/_memray/macho_utils.h

```diff
@@ -0,0 +1,109 @@
+#include <cstring>
+#include <dlfcn.h>
+#include <mach-o/dyld.h>
+#include <mach-o/loader.h>
+#include <mach-o/nlist.h>
+#include <mach/mach.h>
+#include <mach/vm_map.h>
+#include <mach/vm_region.h>
+#include <sys/types.h>
+#include <vector>
+
+#include "hooks.h"
+#include "linker_shenanigans.h"
+#include "logging.h"
+
+#ifndef __APPLE__
+#    error "This file can only be compiled in MacOS systems"
+#endif
+
+#ifdef __LP64__
+typedef struct mach_header_64 mach_header_t;
+typedef struct segment_command_64 segment_command_t;
+typedef struct section_64 section_t;
+typedef struct nlist_64 nlist_t;
+#    define ARCH_LC_SEGMENT LC_SEGMENT_64
+#else
+typedef struct mach_header mach_header_t;
+typedef struct segment_command segment_command_t;
+typedef struct section section_t;
+typedef struct nlist nlist_t;
+#    define ARCH_LC_SEGMENT LC_SEGMENT
+#endif
+
+#ifndef SEG_DATA_CONST
+#    define SEG_DATA_CONST "__DATA_CONST"
+#endif
+
+#ifndef SEG_AUTH_CONST
+#    define SEG_AUTH_CONST "__AUTH_CONST"
+#endif
+
+struct DynamicInfoTable
+{
+    using segment_filter_t = std::function<bool(const char* segname)>;
+
+    const nlist_t* symbol_table;
+    const char* string_table;
+    const uint32_t* dynsym_table;
```

Comment:
These need to be initialized to `nullptr`. As it is, the constructor can return without ever setting them, in which case calling `operator bool` or `getSymbol` would be UB.